server/service/src/document/document_service.rs

```diff
@@ -147,12 +306,37 @@ fn insert_document(
     Ok(doc)
 }
 
+fn delete_document(
+    connection: &StorageConnection,
+    doc: RawDocument,
+) -> Result<(), DocumentDeleteError> {
+    let doc = doc
+        .finalise()
+        .map_err(|err| DocumentDeleteError::InternalError(err))?;
+    let repo = DocumentRepository::new(connection);
+    repo.insert(&doc)?;
+    Ok(())
+}
+
+fn undelete_document(
+    connection: &StorageConnection,
+    doc: RawDocument,
+) -> Result<Document, DocumentUndeleteError> {
+    let doc = doc
+        .finalise()
+        .map_err(|err| DocumentUndeleteError::InternalError(err))?;
+    let repo = DocumentRepository::new(connection);
```

Comment:
You could move the finalise into the generate method and inline the rest, think this would be more in line with how we do it elsewhere